src/estimagic/parameters/nonlinear_constraints.py

```diff
@@ -0,0 +1,468 @@
+from functools import partial
+
+import numpy as np
+import pandas as pd
+from estimagic.differentiation.derivatives import first_derivative
+from estimagic.exceptions import InvalidFunctionError
+from estimagic.parameters.block_trees import block_tree_to_matrix
+from estimagic.parameters.tree_registry import get_registry
+from pybaum import tree_just_flatten
+from pybaum import tree_unflatten
+
+
+CONVERGENCE_ABSOLUTE_CONSTRAINT_TOLERANCE = 1e-5
```

Comment:
I would move this to `algo_options.py` and call it `CONSTRAINTS_ABSOLUTE_TOLERANCE`. The convergence_ prefix is only for convergence criteria of an optimizer, not acceptance of parameter vectors due to constraints.